Allow moving meeting occurrence earlier

oliverguenther opened the pull request:
https://community.openproject.org/work_packages/61424

oliverguenther committed:
Allow moving meeting occurrence earlier

https://community.openproject.org/work_packages/61424

mrmir reviewed (changes requested):
I found some strange behaviour in an edge case:
1. For an occurrence tomorrow
2. Try and change start date to today with a time before the current time -> Get an error
3. Change the start time to a later time than the current time -> Allowed
4. Now with the date already changed to today, change the start time to a time before the current time -> Allowed, which shouldn't be?

myabc commented on modules/meeting/spec/services/meetings/update_service_integration_spec.rb:
```diff
@@ -1,3 +1,32 @@
+# frozen_string_literal: true
+
+#-- copyright
```
@oliverguenther it looks like the copyright notice has been added twice.

myabc commented on modules/meeting/app/contracts/meetings/update_contract.rb:
```diff
@@ -45,14 +48,39 @@ def user_allowed_to_edit
-      return unless model.recurring_meeting_id && model.scheduled_meeting
-      return unless model.changed.intersect?(%w[start_time start_date])
+    def valid_rescheduling_date # rubocop:disable Metrics/AbcSize
+      if model.start_time < Time.zone.now
+        errors.add :start_date, :after_today
+        return
+      end
```
Minor suggestion: ActiveSupport mixes in [`DateAndTime::Calculations`](https://api.rubyonrails.org/classes/DateAndTime/Calculations.html), which gives us `past?`, `future?` and some other sugary methods:

```suggestion
      if model.start_time.past?
        errors.add :start_date, :after_today
        return
      end
```

oliverguenther replied:
I feel like these mixins are less clear than `<` and `<=`, since this clearly communicates if I expect Time.zone.now to be included or not. Granted this isn't a big deal for the application, but it does matter a lot for frozen time in specs and edge cases there.

oliverguenther committed:
Fix case of just changing start_time_hour

oliverguenther commented:
> I found some strange behaviour in an edge case:
> 
> 1. For an occurrence tomorrow
> 2. Try and change start date to today with a time before the current time -> Get an error
> 3. Change the start time to a later time than the current time -> Allowed
> 4. Now with the date already changed to today, change the start time to a time before the current time -> Allowed, which shouldn't be?

This was caused by the `changed` call not including `start_time_hour`. Fixed and added a spec for it.

oliverguenther committed:
Use refind so that instance variables to not leak between specs

mrmir reviewed (approved):
👍
